synapse/rest/client/devices.py

```diff
@@ -80,35 +83,37 @@ def __init__(self, hs: "HomeServer"):
         self.device_handler = hs.get_device_handler()
         self.auth_handler = hs.get_auth_handler()
 
+    class PostBody(RequestBodyModel):
+        auth: Optional[AuthenticationData]
+        devices: List[StrictStr]
+
     @interactive_auth_handler
     async def on_POST(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
         requester = await self.auth.get_user_by_req(request)
 
         try:
-            body = parse_json_object_from_request(request)
+            body = parse_and_validate_json_object_from_request(request, self.PostBody)
         except errors.SynapseError as e:
             if e.errcode == errors.Codes.NOT_JSON:
-                # DELETE
+                # TODO: Can/should we remove this fallback now?
                 # deal with older clients which didn't pass a JSON dict
                 # the same as those that pass an empty dict
-                body = {}
+                body = self.PostBody.parse_obj({})
             else:
                 raise e
 
-        assert_params_in_dict(body, ["devices"])
-
         await self.auth_handler.validate_user_via_ui_auth(
             requester,
             request,
-            body,
+            body.dict(exclude_unset=True),
```

Comment:
any interest in making `validate_user_via_ui_auth` use Pydantic gizmos rather than taking a dict in the near future?

Reply:
Maybe. UI auth is scary, and it works on free-form fields so I'm not sure we can do much better than a dict.

It'd be good to pass in the model directly though, so each caller doesn't have to worry about the exclude_unset footgun.

FWIW my rough plan is to get the models in use at the rest layer as a first pass, and then gradually make more use of them in the handler layers over time.